lib/matplotlib/__init__.py

```diff
@@ -1515,6 +1515,71 @@ def test(verbosity=1):
 
 test.__test__ = False  # nose: this function is not a test
 
+
+def _replacer(data, key):
+    # if key isn't a string don't bother
+    if not isinstance(key, six.string_types):
+        return key
+    # try to use __getitem__
+    try:
+        return data[key]
+    # key does not exist, silently fall back to key
+    except KeyError:
+        return key
+
+
+def unpack_labeled_data(wl_args=None, wl_kwargs=None, label_pos=None):
+    """
+    A decorator to add a 'data' kwarg to any a function.  The signature
+    of the input function must be ::
+
+       def foo(ax, *args, **kwargs)
+
+    so this is suitable for use with Axes methods.
+    """
+    if label_pos is not None:
+        label_arg, label_kwarg = label_pos
+
+    if wl_kwargs is not None:
+        wl_kwargs = set(wl_kwargs)
+    if wl_args is not None:
+        wl_args = set(wl_args)
+
+    def param(func):
+        @functools.wraps(func)
+        def inner(ax, *args, **kwargs):
+            data = kwargs.pop('data', None)
+            if data is not None:
+                if wl_args is None:
+                    args = tuple(_replacer(data, a) for a in args)
+                else:
+                    args = tuple(_replacer(data, a) if j in wl_args else a
+                                 for j, a in enumerate(args))
+
+                if wl_kwargs is None:
+                    kwargs = dict((k, _replacer(data, v))
+                                  for k, v in six.iteritems(kwargs))
+                else:
+                    kwargs = dict(
+                        (k, _replacer(data, v) if k in wl_kwargs else v)
+                        for k, v in six.iteritems(kwargs))
+            if (label_pos is not None and ('label' not in kwargs or
+                    kwargs['label'] is None)):
+                if len(args) > label_arg:
+                    try:
+                        kwargs['label'] = args[label_arg].name
```

Comment:
Why not use @mwaskom's suggestion of using the text label instead of the `.name` attribute? That seems safer:

> To be clear I'm not saying matplotlib needs to extract a .name attribute from a vector, just that if semantic names are used to draw the plot, they should end up as labels too.

`.name` will also work with xray, but the smaller we can make the labeled data spec, the better.